public/js/utils/source-map.js

```diff
@@ -1,160 +1,206 @@
+  // If it has already been requested, return the request. An
+  // important behavior here is that if it's in the middle of
+  // requesting it, all subsequent calls will block on the initial
+  // request.
+  const existingRequest = sourceMapRequests.get(generatedSource.id);
+  if (existingRequest) {
+    return existingRequest;
+  } else if (!generatedSource.sourceMapURL || !isEnabled("sourceMaps")) {
+    return Promise.resolve(null);
   }
 
-  const generatedSourceId = getGeneratedSourceId(source);
-  const originalSource = getSource(state, generatedSourceId);
+  async function request() {
+    // Fetch the sourcemap over the network and create it.
+    const sourceMapURL = _resolveSourceMapURL(generatedSource);
+    const fetched = await networkRequest(
+      sourceMapURL, { loadFromCache: false }
+    );
 
-  if (originalSource) {
-    return originalSource.toJS();
+    // Create the source map and fix it up.
+    const map = new SourceMapConsumer(fetched.content);
+    _setSourceMapRoot(map, sourceMapURL, generatedSource);
+    return map;
   }
 
-  return source;
+  // Fire off the request, set it in the cache, and return it.
+  const req = request();
+  sourceMapRequests.set(generatedSource.id, req);
+  return req;
+}
+
+function getSourceMap(generatedSourceId) {
+  return sourceMapRequests.get(generatedSourceId);
 }
 
-async function getGeneratedLocation(state: AppState, location: Location) {
-  const source: any = getSource(state, location.sourceId);
+function installSourceMappings(generatedId, url, code, mappings) {
+  const generator = new SourceMapGenerator({ file: url });
+  mappings.forEach(mapping => generator.addMapping(mapping));
+  generator.setSourceContent(url, code);
 
-  if (!source) {
+  const map = SourceMapConsumer(generator.toJSON());
+  sourceMapRequests.set(generatedId, Promise.resolve(map));
+}
+
+async function getGeneratedLocation(location: Location, state) {
+  if (!isOriginalId(location.sourceId)) {
     return location;
   }
 
-  if (await isOriginal(source.toJS())) {
-    return await getGeneratedSourceLocation(source.toJS(), location);
+  const originalSource = getSource(state, location.sourceId).toJS();
+  const generatedSourceId = originalToGeneratedId(location.sourceId);
+  const map = await getSourceMap(generatedSourceId);
+  if (!map) {
+    return location;
   }
 
-  return location;
+  const { line, column } = map.generatedPositionFor({
+    source: originalSource.url,
+    line: location.line,
+    column: location.column == null ? 0 : location.column
+  });
+
+  return {
+    sourceId: generatedSourceId,
+    line: line,
+    // Treat 0 as no column so that line breakpoints work correctly.
+    column: column === 0 ? undefined : column
+  };
 }
 
-async function getOriginalLocation(state: AppState, location: Location) {
-  const source: any = getSource(state, location.sourceId);
+async function getOriginalLocation(location: Location) {
+  assert(isGeneratedId(location.sourceId),
```

Comment:
hmm, i could be wrong here, but we might want to handle original sources by just returning the location. I'll have to think about this a bit.